File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        st=0
        l=len(inorder)
        def bt(i,f,pre,indorder):
            nonlocal st,l
            if i<=f and st<l:
                ind=inorder.index(pre[st])
                new=TreeNode(inorder[ind])
                st+=1
                new.left=bt(i,ind-1,pre,inorder)
                new.right=bt(ind+1,f,pre,inorder)
                return new
            else:
                return None
        return bt(0,len(preorder)-1,preorder, inorder)

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int preorder[],inorder[],n,i=0;

    public int findIndex(int x){
        int ind=-1;
        for(int i=0;i<n;i++)
            if(inorder[i]==x)
                return i;
        return -1;
    }

    public TreeNode find(int j, int k){
        if(k-j+1<=0||i>=n)  return null;
        int ind=findIndex(preorder[i]);
        TreeNode root = new TreeNode(preorder[i++]);
        root.left=find(j,ind-1);
        root.right=find(ind+1,k);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder=preorder;
        this.inorder=inorder;
        n=inorder.length;
        return find(0,n-1);
    }
}